### ISSUE TITLE
Toggle files browse between fancy and grid list

### ISSUE THREAD

**Issue by bog-dan-ro** [bog-dan-ro/spectacol#18]
http://www.worldofspectrum.org/forums/discussion/comment/855009#Comment_855009

**Comment by ghost** [comment]
When you implement this can you also make the games list remember where I got to in the list. If you have a lot of games you have to keep scrolling through from the beginning each time. Also annoying if you accidently push up while scrolling through and it changes directory (in the cover flow)!

EDIT. I'll add this as a seperate issue.

**Comment by ghost** [comment]
How's it going ?

**Comment by bog-dan-ro** [comment]
Not well, I was in a little vacation and I didn't had time to check it yet.

**Comment by ghost** [comment]
The 'How's it going?' was meant for 'How's it going with the new grid list.' I didn't know you were on vacation. Hope you had a good time.

**Comment by bog-dan-ro** [comment]
Well, the grid list is already implemented, I'm already using it for online search. 
The only problem is my free time which currently is quite limited ...

**Comment by ghost** [comment]
Any chance you could post a new Beta with the new file list, I can't get on with that Cover Flow view ?

**Comment by bog-dan-ro** [comment]
I'll try to do something this weekend, but it's not a promise!

**Comment by ghost** [comment]
There seems to be a lot of pointless stuff requested on this emulator that will probably take up a lot of your time and IMO shouldn't even be bothered with, like Voice Control, Webcam Control, even the Metadata thing could be left till last. I think the emulator should at least try to equal whats currently available first (otherwise, what's the point when other emulators can already do it). Things like being able to clearly select a game from a list, optional original loading speed (which is a huge part of Spectrum nostalgia), a nice view while playing the games (with no onscreen icons especially when using external controls) etc. There still isn't one emulator on Android that can do it all. Spectaculator  is probably the best by a long shot but it can't load SCL & TRD. The best Spectrum emulator I have ever used is still the Nintendo DS emulator ZXDS. It does everything. I'm hoping this emulator can get there (or unless Spectaculator adds SCL support first). :)

**Comment by bog-dan-ro** [comment]
Most of these requests (Voice Control, Webcam Control) are postpone for future release and they will not get in 2.0.

I started a new thread here http://www.worldofspectrum.org/forums/discussion/52504/spectacol-2-0-new-features to see which features are most wanted for 2.0. Some of your requests are already on my list, but you need to be patient :) 

The metadata it is something that personally I care a lot.

Regarding the other Android emulators, well, that was the reason why I choose to create Spectacol, because none of them didn't impressed me. I tried Spectaculator (I didn't knew it until you told me about it) and it didn't impressed me either.

On **huge** advantage of Spectacol over the other emulators is that it's **free**! This means that anyone can contribute or even fork it. This mean you don't need to wait for me to implement any feature you prefer, if you have some C++ knowledges (QML language is very simple) you can implement them yourself.

Last but not least, as long as I'll run this project you'll not be bother you with any annoying ads or messages that asks donations.

**Comment by ghost** [comment]
Spectaculator didn't impress you ? I take it you only tried the free version which only lets you play the preinstalled games and has preset options ? Nearly every feature I wish Spectacol had (It also has by far the best RZX compatibility). USP's only benefit is autobooting SCL files but it seems to have some stupid timing that plays some SCL games too fast. I also recently tried the emu 'Speccy' which has some nice features. Currently I have to have all these apps to have a decent Spectrum experience. I hope Spectacol can combine the stand out features from all three but at the moment it's at the back of the queue unfortunately.

**Comment by ghost** [comment]
Just out of interest, how come the new beta is still only using the cover flow view ? I thought you said (above) that it was already implemented ?

**Comment by ZxAsSeMbLeR** [comment]
Maybe this is not the best place to speak about emus comparison (inside a issue), but as there have been already opinions speaking about other ZX Spectrum emulators for Android in this thread, which I have installed, I will add my point of view about the pros and cons of the (for me) best Zx spectrum emus for Android:
- Speccy Deluxe
- USP
- Spectaculator
- Spectacol

Speccy Deluxe -->
Pros: A very, very, very big amount of customization options. This is one of its strong points. You can find a lot of useful options, filters, controls, virtual gamepad layout editor, tape browser, etc. It has also the possibility to see a cover or screenshot of each game if you put the corresponding files in the same subfolder as the game itself. It has also Online Game Browsing. The display options are very good and the control is also good. Frequently updated.

Cons: No zip browsing (but it supports zipped files), This has the worst 48k emulated sound of the group I am reviewing. It can't load some protected turbo games in TZX format (that is really bad, because those games load ok in the rest of emulators). Speccy Deluxe is a paid version (there is also a free one with ads and some options removed).

USP --> 
Pros: One of the best nowadays. The sound is very,very good. The image is also nearly perfect. It has zip browsing support and loads nearly everything with a fast forward button to accelerate the game execution or to enjoy real time loading but in less time if you have the button pressed. Good control and Online Game Browsing. It's free.

Cons: Too little options. Not very customizable. The stretching screen option still has the ZX Spectrum borders in both sides. The change between keyboard/joystick is strange and has problems in some devices with no back, home Android buttons directly to use.

Spectaculator --> 

Pros: Great sound (but sometimes depending on the device there is is a kind of "echo" and you have to leave the game and enter again to fix this) and great image quality. The control is also really good. All games and formats are emulated really well. 

Cons: No updates in last year. You can't browse all your collection of games. You have to add the games only in a folder (without subfolders), so you have to select and copy only your prefered games to use with this emu (if the number of games is big the emulator takes a lot of time to load because it stores the configuration for each game individually). No tape browsing, and no fast forward option. No Online Game Browsing. Is is not free if you want to play your own games.

Spectacol -->

Pros: Very promising, free and still in development and with the developer listening our suggestions ;-). The sound is really good and the image quality is also good after last beta revisions (stretching, filters, smoothing...). I really like the new browsing style. It's really nice to have screenshots of all the games as I browse through my collection of games in folders/subfolders. That if, I use Spectacol specially with games in z80 format and uncompressed to really enjoy the browsing style (I have made a subfolder with all my thousands of games in z80 just to use with this emu). The control is also very good. It has also Online Game Browsing.

Cons: 

A lot of the cons are already in the list of enhancement/bug list as pending things, so if part of them
are covered this emu could be the best: Zip support, browsing tapes, Getting metada from TZX, Pokes...

Well, so my perfect emu would be Spectacol with features of USP, some customization options from Speccy Deluxe mixed with the ingame quality of Spectaculator ;-). I have the four installed and the ones I use more frequently are Spectacol and USP. I use Speccy Deluxe if I want to see the covers and screenshots of my collection of games when browsing (another copy of my collection adapted just to use in that emu), and I have been leaving apart Spectaculator (which I really like once you are inside a game) but is really bad for browsing your collection of games before starting one of them. Some months ago Spectaculator and Speccy where the emus I used more now replaced by 
Spectacol and USP for the reasons I mentioned above.

**Comment by ghost** [comment]
Nice comparison. I would also add that Speccy has the worst compatibility of any Spectrum emulator, hardly anything I tried would work but it does have some excellent options like choice of 48k/128k on-screen keyboard that's resizable. I use Spectaculator for RZX replays as it is by far the most compatible. USP seems to have some weird timing issues, some games and RZX files play at the wrong speed !! Especially try the RZX for IK+ and the music keeps speeding up and slowing down ! (Also happens with certain games). This ruined USP for me. One thing that all these emulators have in common (and the issue of this thread) is they have a basic vertical file list where you can easily read the filenames and can see the next few or previous filenames easily.

**Comment by bog-dan-ro** [comment]
@windale: I tried to just use the grid list for local files but I it was not that easy as I expected. When I'll finish it I'll close this issue ;-)

@ZxAsSeMbLeR: thanks for comparison

**Comment by ZxAsSeMbLeR** [comment]
Good annotation for the RZX files. I hardly use them, that's the reason why I like USP. It's good to know Spectaculator supports them perfectly ;-)

**Comment by ghost** [comment]
3.0 !! Gutted. I take it everyone is a big fan of this iTunes Cover Flow view and are probably only using z80 files.

**Comment by bog-dan-ro** [comment]
I had to choose the most important two issues from #28, #6 and this one to go to 2.0, and I choose first two.
I'll start working on 3.0 soon after 2.0 is shipped. I'll also add only a few features to every release to have faster releases...

**Comment by ghost** [comment]
OK, but I wouldn't say ZIP support was important, Spectrum files are tiny anyway and you barely save anything when zipping. It doesn't take long to unzip a collection of files (TOSEC). People don't want to bother unzipping because the collection they downloaded was already zipped ?

The thing that confuses me is that you said the following in a post above : 
"Well, the grid list is already implemented, I'm already using it for online search."

**Comment by ghost** [comment]
I just tried your online browser, I've never used it before because I already have an offline collection. With this list you can only see one game per page because of the big screenshot. I hope when you implement the local file list it will be something like Spectaculator/USP with lots of files visible and a way to fast scroll through.
![screenshot_2016-05-21-16-36-03](https://cloud.githubusercontent.com/assets/12862298/15449348/eefa4174-1f72-11e6-9ba3-4588f74c6a21.png)

I think with something this size you could still add the screenshot. Thanks.

**Comment by bog-dan-ro** [comment]
There are quite a lot people asking for zip support, personally I hate it a lot but I must do it because is one of the most wanted features.

The online browser shows a grid of files and it should fit much more than one per page (usually ~9). If not then there is a bug and a screen will help to to see what's wrong and to fix it.

Back to this issue, I'd like to let you decide how many and in witch way you want to see your files:
- in a cover flow (well in this mode you want be able to decide the size of how many you'll see)
- in a grid, and you can choose how large will be the icons, so they will fit more or less on your screen.
- in a list, and again you can choose how large will be the icons

But to implement all these things will take time (much more than I anticipated) and it will delay too much the 2.0 release

**Comment by ghost** [comment]
This is what I see when using the online browser, as you can see, only 1 file visible per page/screen.

![screenshot_2016-05-21-17-43-16](https://cloud.githubusercontent.com/assets/12862298/15449663/b9873872-1f7b-11e6-803d-1684dc9058f6.png)

I only have a 5 inch screen on my device. Please do not add any of your time fixing this online browser just for me because I don't use it anyway.

(If when implementing the local file selector, the basic list will always get my vote if you were adding the choices one at a time).

**Comment by bog-dan-ro** [comment]
It's not just for your :) It's for everybody who use it , and from what I can see is super ugly ....

**Comment by ghost** [comment]
I assumed it worked fine on everyone else's because nobody else reported anything.
